bigchaindb/core.py

```diff
@@ -217,26 +217,26 @@ def get_blocks_status_containing_tx(self, txid):
         else:
             return None
 
-    def get_tx_by_payload_hash(self, payload_hash):
+    def get_tx_by_payload_uuid(self, payload_uuid):
         """Retrieves transactions related to a digital asset.
 
         When creating a transaction one of the optional arguments is the `payload`. The payload is a generic
         dict that contains information about the digital asset.
 
-        To make it easy to query the bigchain for that digital asset we create a sha3-256 hash of the
-        serialized payload and store it with the transaction. This makes it easy for developers to keep track
-        of their digital assets in bigchain.
+        To make it easy to query the bigchain for that digital asset we create a UUID for the payload and 
+        store it with the transaction. This makes it easy for developers to keep track of their digital 
+        assets in bigchain.
 
         Args:
-            payload_hash (str): sha3-256 hash of the serialized payload.
+            payload_uuid (str): the UUID for this particular payload.
 
         Returns:
             A list of transactions containing that payload. If no transaction exists with that payload it
             returns an empty list `[]`
         """
 
         cursor = r.table('bigchain') \
-            .get_all(payload_hash, index='payload_hash') \
+            .get_all(payload_uuid, index='payload_uuid') \
```

Comment:
Ok, we have another problem here. The secondary index will return all matching documents in the table and the documents in the bigchain table are blocks not transactions. So this will return the block where the matching transaction exists.

We need to improve the query a bit to return only the matching transactions:
1. Once we get the matching block(s) we need to flatten all the transactions on those blocks
2. Filter the transactions based on the criteria

```python
cursor = r.table('bigchain') \
               .get_all(payload_uuid, index='payload_uuid') \
               .concat_map(lambda block: block['block']['transactions']) \  # concat all transaction lists
               .filter(lambda transaction: transaction['transaction']['data']['uuid'] == payload_uuid) \
               .run(b.conn)
```